Content.Server/Construction/PartExchangerSystem.cs

```diff
@@ -35,30 +35,83 @@ private void OnDoAfter(EntityUid uid, PartExchangerComponent component, DoAfterE
         if (args.Cancelled || args.Handled || args.Args.Target == null)
             return;
 
-        if (!TryComp<MachineComponent>(args.Args.Target.Value, out var machine))
-            return;
-
         if (!TryComp<ServerStorageComponent>(uid, out var storage) || storage.Storage == null)
             return; //the parts are stored in here
 
+        var machineParts = new List<MachinePartComponent>();
+
+        foreach (var ent in storage.Storage.ContainedEntities) //get parts in RPED
+        {
+            if (TryComp<MachinePartComponent>(ent, out var part))
+                machineParts.Add(part);
+        }
+
+        TryExchangeMachineParts(args.Args.Target.Value, storage, machineParts);
+        TryConstructMachineParts(args.Args.Target.Value, storage, machineParts);
+
+        args.Handled = true;
+    }
+
+    private void TryExchangeMachineParts(EntityUid uid, ServerStorageComponent storage, List<MachinePartComponent> machineParts)
+    {
+        if (!TryComp<MachineComponent>(uid, out var machine) || storage.Storage == null)
+            return;
+
         var board = machine.BoardContainer.ContainedEntities.FirstOrNull();
 
         if (board == null || !TryComp<MachineBoardComponent>(board, out var macBoardComp))
             return;
 
-        var machineParts = new List<MachinePartComponent>();
-
-        foreach (var ent in storage.Storage.ContainedEntities) //get parts in RPED
+        foreach (var ent in new List<EntityUid>(machine.PartContainer.ContainedEntities)) //clone so don't modify during enumeration
```

Comment:
```suggestion
        foreach (var ent in new ValueList<EntityUid>(machine.PartContainer.ContainedEntities)) //clone so don't modify during enumeration
```

At least if it takes a ctor.